### DISCUSSION TITLE
import error for tensorflow and keras

### DISCUSSION THREAD

**Discussion by wenchaoj** [ylikx/forpy#45] [Q&A]
Hi,

Thank you very much for providing the forpy library! I am hoping to link Fortran and Python in order to utilize neural net tools in Python during the execution of a larger Fortran program. I have worked through your online examples and am trying to build upon them for my project. I am stuck at the point of importing tensorflow and keras. 

Below are my codes where I try to import various Python modules. 

```
program importexample
use forpy_mod
implicit none

integer :: ierror
type(module_py) :: opt,math,tf,keras,datetime

ierror = forpy_initialize()
ierror = import_py(opt, "scipy.optimize")
write(*,*) "scipy.optimize imported return ierror", ierror
ierror = import_py(keras, "keras.model")
write(*,*) "keras.model imported return ierror", ierror
ierror = import_py(math, "numpy")
write(*,*) "numpy imported return ierror", ierror
ierror = import_py(tf, "tensorflow")
write(*,*) "tensoflow imported return ierror", ierror
ierror = import_py(datetime, "datetime")
write(*,*) "datetime imported return ierror", ierror
call forpy_finalize()

end program
```

After I compile with ifort importexample.F90 forpy_mod.o `python3-config --ldflags --embed` and execute with ./a.out, I find the ierror=0 for scipy.optimize, numpy and datetime; but ierror=1 for keras.model and tensorflow. All the

I had no problem implementing any example program that you provided on this repository. I am using python3 on a university-provided computing platform:
[uctpwji@login14 Python]$ which python
/shared/ucl/apps/python/bundles/python39-6.0.0/venv/bin/python
[uctpwji@login14 Python]$ python3-config --ldflags
 -L/lustre/shared/ucl/apps/python/3.9.10/gnu-4.9.2/lib  -lcrypt -lpthread -ldl  -lutil -lm -lm 

Any tips or suggestions will be very appreciated! Many thanks again for making Forpy available!

**Comment by ylikx** [comment]
Hi,
insert a `call err_print` statement after the import to get a more helpful error message. I hope this helps.

**Reply by wenchaoj** [reply]
Thank you so much for your suggestion! I realized that I didn't load the correct modules on the university's computing platform before running the program with Forpy. After addressing that, the program that I posted before works fine. 
Now I am stuck at another point. I am trying to call a tensorflow function and cast the output array back to Fortran type. I get a compilation error that says "There is no matching specific function for this generic function reference.   [CAST]"
Do you think you can provide some example codes that calls a Python function, taking Fortran arrays as inputs and generating array outputs, and then casts the output back to Fortran arrays?
I see your wonderful example program "scipy_example01", which is doing to most of the above. What I cannot figure out myself is: when the output of the python function doesn't have 'attributes', when the output is simply an array or tensor, how should we get the output back in Fortran type?
Thank you so much for your guidance!

**Comment by ylikx** [comment]
I guess you probably want to call a method on maybe a Tensorflow tensor, but you only have a generic forpy object? 
You can also use call_py for that, for example:
ierror = call_py(retval, tensorflow_object, "numpy") 
to call the "numpy" method of a tensorflow tensor.
